roachtest: stop using ClusterSpec.Cloud in test code

RaduBerinde opened the pull request:
In this PR we fix all test registration code which was depending on the flags passed to `roachtest` (indirectly, through TestSpec). In the process, I tried to clean up various things I had to touch.

This makes progress in the direction of #104029 and of making the registry and `TestSpec` not depend at all on the flags (the flags should come in later, when we create an actual cluster from the spec).

#### roachtest: spec: simplify Option

This commit reduces boilerplate in the `Option` code by making it a
func instead of an interface with an apply func. This way each option
can define the function inline instead of having to define a type.

Epic: none
Release note: None

#### roachtest: spec: provide options for GCE/AWS settings

This commit moves GCE and AWS specific settings to their own
inline structs and adds Options for them.

Epic: none
Release note: None

#### roachtest: clean up instance type specification

`ClusterSpec` now provides options for GCE and AWS machine types.

Epic: none
Release note: None

#### roachtest: clean up zones specification

`ClusterSpec` now provides options for GCE and AWS zones
specification.

Epic: none
Release note: None

#### roachtest: stop using ClusterSpec.Cloud in test code

This change removes all remaining uses of `ClusterSpec.Cloud` except
those internal to roachtest. Code that is part of running a test now
uses `Cluster.Cloud()` instead.

Informs: #104029
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/111324)
<!-- Reviewable:end -->

RaduBerinde commented:
Updated (resolved conflicts).

renatolabs reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111324#-:-Nfpu6cNAe7OlAWT7lIv:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @RaduBerinde, @smg260, and @srosenberg)

___
*[`build/release/teamcity-mark-build.sh` line 31 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/111324#-Nfpu3cl2TDJD2SB40rW:-Nfpu3cm7_jv363XSD_k:b-fib7sn) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed5bb3e3b8b47fc0d2733d870e1bb41f2bd6da08/build/release/teamcity-mark-build.sh#L31)):*
> ```Shell
> 
>   log_into_gcloud
>   gawscloud container images add-tag "${gcr_repository}:${TC_BUILD_BRANCH}" "${gcr_repository}:latest-${release_branch}-${build_label}-build"
> ```

This doesn't look right?


<!-- Sent from Reviewable.io -->

smg260 reviewed (commented):
Are all the `OnlyGCE` actually only GCE? At least one example that worked locally is `ledger/nodes=6/multi-az`, whether intended to, or not.

> Reviewed 8 of 18 files at r4, 20 of 28 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111324#-:-Nfq1zSA73fSga-8vPyE:b-koi61z)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @RaduBerinde and @srosenberg)



<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
TFTRs!!

It's a multi AZ test, feels like there's no point to run locally (other than perhaps debugging the test). Note that you will always be able to run whatever you want by adding a flag to override the advertised compatibilities.

I will wait for #111633 to go in and address any conflicts.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111324#-:-Nfq7Sa2AWwtkSQsFy3y:b6cjhzz)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @renatolabs, @smg260, and @srosenberg)

___
*[`build/release/teamcity-mark-build.sh` line 31 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/111324#-Nfpu3cl2TDJD2SB40rW:-Nfq8EZR-bBYG0k8M6lO:b26emnd) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed5bb3e3b8b47fc0d2733d870e1bb41f2bd6da08/build/release/teamcity-mark-build.sh#L31)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

This doesn't look right?
</blockquote></details>

Wow.. fat fingers. Fixed.


<!-- Sent from Reviewable.io -->

smg260 reviewed (approved):
> Reviewed 1 of 38 files at r6, 37 of 37 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111324)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @renatolabs and @srosenberg)



<!-- Sent from Reviewable.io -->

RaduBerinde committed:
roachtest: spec: simplify Option

This commit reduces boilerplate in the `Option` code by making it a
func instead of an interface with an apply func. This way each option
can define the function inline instead of having to define a type.

Epic: none
Release note: None

RaduBerinde committed:
roachtest: spec: provide options for GCE/AWS settings

This commit moves GCE and AWS specific settings to their own
inline structs and adds Options for them.

Epic: none
Release note: None

RaduBerinde committed:
roachtest: clean up instance type specification

`ClusterSpec` now provides options for GCE and AWS machine types.

Epic: none
Release note: None

RaduBerinde committed:
roachtest: stop using ClusterSpec.Cloud in test code

This change removes all remaining uses of `ClusterSpec.Cloud` except
those internal to roachtest. Code that is part of running a test now
uses `Cluster.Cloud()` instead.

Informs: #104029
Release note: None

RaduBerinde committed:
roachtest: clean up zones specification

`ClusterSpec` now provides options for GCE and AWS zones
specification.

Epic: none
Release note: None

RaduBerinde commented:
Test build before merging: https://teamcity.cockroachdb.com/viewLog.html?buildId=12056487&buildTypeId=Cockroach_Nightlies_RoachtestNightlyGceBazel&tab=buildResultsDiv&branch_Cockroach_Nightlies=111251

RaduBerinde commented:
bors r+
